dataRetriever/dataPool/headersCache/headersNonceCache.go

```diff
@@ -0,0 +1,252 @@
+	if !ok {
+		return
+	}
+
+	//remove header from header list
+	for index, headerD := range hdrListD.headerList {
+		if !bytes.Equal(headerD.headerHash, headerHash) {
+			continue
+		}
+
+		hdrListD.headerList = append(hdrListD.headerList[:index], hdrListD.headerList[index+1:]...)
+		hnc.hdrsCounter.decrement(hdrInfo.headerShardId)
+
+		if len(hdrListD.headerList) == 0 {
+			hnc.hdrNonceCache[hdrInfo.headerShardId].removeElement(hdrInfo.headerNonce)
+			return
+		}
+
+		hnc.hdrNonceCache[hdrInfo.headerShardId].addElement(hdrInfo.headerNonce, hdrListD)
+		return
+	}
+}
+
+func (hnc *headersNonceCache) removeHeaderNonceByNonceAndShardId(hdrNonce uint64, shardId uint32) int {
+	_, ok := hnc.hdrNonceCache[shardId]
+	if !ok {
+		return 0
+	}
+
+	hdrListD, ok := hnc.hdrNonceCache[shardId].getHeadersDetailsListFromSMap(hdrNonce)
+	if !ok {
+		return 0
+	}
+
+	hdrsHashes := make([][]byte, 0)
+	for _, hdrDetails := range hdrListD.headerList {
+		hdrsHashes = append(hdrsHashes, hdrDetails.headerHash)
+		hnc.hdrsCounter.decrement(shardId)
+	}
+
+	hnc.hdrNonceCache[shardId].removeElement(hdrNonce)
+
+	//remove elements from hashes map
+	for _, hash := range hdrsHashes {
+		hnc.headersByHash.deleteElement(hash)
+	}
+
+	return len(hdrsHashes)
+}
+
+func (hnc *headersNonceCache) getHeadersByNonceAndShardId(hdrNonce uint64, shardId uint32) ([]headerDetails, bool) {
+	hnc.mutHeadersMap.RLock()
+	defer hnc.mutHeadersMap.RUnlock()
+
+	_, ok := hnc.hdrNonceCache[shardId]
+	if !ok {
+		return nil, false
+	}
+
+	headersList, ok := hnc.hdrNonceCache[shardId].getHeadersDetailsListFromSMap(hdrNonce)
+	if !ok {
+		return nil, false
+	}
+
+	return headersList.headerList, true
+}
+
+func (hnc *headersNonceCache) getNumHeaderFromCache(shardId uint32) int64 {
+	return hnc.hdrsCounter.getNumHeaderFromCache(shardId)
+}
+
+func (hnc *headersNonceCache) lruEviction(shardId uint32) {
+	nonces := hnc.hdrNonceCache[shardId].getNoncesTimestampSorted()
+
+	var numHashes int
+	for i := 0; i < hnc.numHeadersToRemove && i < len(nonces); i++ {
+		numHashes += hnc.removeHeaderNonceByNonceAndShardId(nonces[i], shardId)
+
+		if numHashes >= hnc.numHeadersToRemove {
+			break
+		}
+	}
+}
+
+func (hnc *headersNonceCache) removeHeaderByHash(hash []byte) {
+	hnc.mutHeadersMap.Lock()
+	defer hnc.mutHeadersMap.Unlock()
+
+	info, ok := hnc.headersByHash.getElement(hash)
+	if !ok {
+		return
+	}
+
+	//remove header from first map
+	hnc.removeHeaderNonceCache(info, hash)
+	//remove header from second map
+	hnc.headersByHash.deleteElement(hash)
+}
+
+func (hnc *headersNonceCache) getHeaderByHash(hash []byte) (data.HeaderHandler, error) {
+	info, ok := hnc.headersByHash.getElement(hash)
+	if !ok {
+		return nil, ErrHeaderNotFound
+	}
+
+	headersList, ok := hnc.getHeadersByNonceAndShardId(info.headerNonce, info.headerShardId)
+	if !ok {
+		return nil, ErrHeaderNotFound
+	}
+
+	for _, hdrDetails := range headersList {
+		if bytes.Equal(hash, hdrDetails.headerHash) {
```

Comment:
Add extra function `findHeader` in `headerListDetails` perhaps? For encapsulation, I mean. Not sure if you can without some other refactor to `getHeadersByNonceAndShardId`.